syntax = "proto3";

package sift.assets.v1;

import "google/api/annotations.proto";
import "google/api/field_behavior.proto";
import "google/protobuf/timestamp.proto";
import "protoc-gen-openapiv2/options/annotations.proto";


option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_swagger) = {
  info: {title: "Asset Service"}
};

service AssetService {
  option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_tag) = {
    description: "Service to programmatically interact with [assets](/glossary#asset)."
    external_docs: {
      url: "/data-model",
      description: "Read more about what assets are.";
    }
  };

  // Deletes an asset.
  rpc DeleteAsset(DeleteAssetRequest) returns (DeleteAssetResponse) {
    option (google.api.http) = {delete: "/api/v1/assets/{asset_id}"};
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "DeleteAsset"
      description: "Deletes an asset."
    };
  }

  // Retrieves an asset.
  rpc GetAsset(GetAssetRequest) returns (GetAssetResponse) {
    option (google.api.http) = {get: "/api/v1/assets/{asset_id}"};
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "GetAsset"
      description: "Retrieves an asset."
    };
  }

  // Retrieves assets using an optional filter.
  rpc ListAssets(ListAssetsRequest) returns (ListAssetsResponse) {
    option (google.api.http) = {get: "/api/v1/assets"};
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "ListAssets"
      description: "Retrieves assets using an optional filter."
    };
  }
}

message Asset {
  string asset_id = 1 [(google.api.field_behavior) = REQUIRED];
  string name = 2 [(google.api.field_behavior) = REQUIRED];
  string organization_id = 4 [(google.api.field_behavior) = REQUIRED];
  google.protobuf.Timestamp created_date = 5 [(google.api.field_behavior) = REQUIRED];
  string created_by_user_id = 6 [(google.api.field_behavior) = REQUIRED];
  google.protobuf.Timestamp modified_date = 7 [(google.api.field_behavior) = REQUIRED];
  string modified_by_user_id = 8 [(google.api.field_behavior) = REQUIRED];
}

// The request for a call to `AssetService_ListAssets`.
message ListAssetsRequest {
  // The maximum number of assets to return.
  // The service may return fewer than this value.
  // If unspecified, at most 50 assets will be returned.
  // The maximum value is 1000; values above 1000 will be coerced to 1000.
  uint32 page_size = 1 [(google.api.field_behavior) = OPTIONAL];

  // A page token, received from a previous `ListAssets` call.
  // Provide this to retrieve the subsequent page.
  // When paginating, all other parameters provided to `ListAssets` must match
  // the call that provided the page token.
  string page_token = 2 [(google.api.field_behavior) = OPTIONAL];

  // A [Common Expression Language (CEL)](https://github.com/google/cel-spec) filter string.
  // Available fields to filter by are `asset_id`, `created_by_user_id`, `modified_by_user_id`,
  // `created_date`, `modified_date`, and `name`.
  // For further information about how to use CELs, please refer to [this guide](https://github.com/google/cel-spec/blob/master/doc/langdef.md#standard-definitions).
  // For more information about the fields used for filtering, please refer to [this definition](/protocol-buffers/documentation#asset). Optional.
  string filter = 3 [(google.api.field_behavior) = OPTIONAL];
}

// The result of a call to `AssetService_ListAssets`.
message ListAssetsResponse {
  repeated Asset assets = 1;
  string next_page_token = 5;
}

// The request for a call to `AssetService_DeleteAsset` to delete a single existing annotation by its asset_id.
message DeleteAssetRequest {
  // The id of the asset to be deleted. Required.
  string asset_id = 1 [(google.api.field_behavior) = REQUIRED];
}

// The response of a call to `AssetService_DeleteAsset`.
message DeleteAssetResponse {}

// The request for a call to `AssetService_GetAsset` to retrieve a single existing asset by its asset_id.
message GetAssetRequest {
  // The id of the asset to be retrieved. Required.
  string asset_id = 1 [(google.api.field_behavior) = REQUIRED];
}

message GetAssetResponse {
  Asset asset = 1;
}
